### PR TITLE
feat: Handle ACIR calls in the debugger

### LINKED ISSUE #4824: Debugger: Handle multiple ACIR calls
### Problem

With the introduction of https://github.com/noir-lang/noir/issues/4428 we now support multiple ACIR function calls. Inlining remains the default, however, functions marked with `#[fold]` (or potentially other inline markers in the future) will be treated as separate ACIR functions. Execution has been updated in the ACVM and nargo to handle these new non-inlined methods but not in the debugger. 

The relevant work that updated the codegen and execution for non-inlined ACIR can be found linked in the main issue linked above.

cc @ggiraldez @mverzilli

### Happy Case

It would be nice to be able to use the debugger even when executing non-inlined ACIR. This will be especially useful as we potentially add more inline markers such as `#[inline(never)]`. 

### Project Impact

Nice-to-have

### Impact Context

_No response_

### Workaround

None

### Workaround Description

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by ggiraldez**
# Description

## Problem\*

Resolves #4824

With the introduction of ACIR calls to allow non-inlined ACIR functions, the debugger did not have proper support for programs that had them, neither when compiling in ACIR mode, nor in the default mode which compiles the full program to Brillig.

## Summary\*

This PR depends on #4941 and is built on top of it.

The changes allow compiling a program with `#[fold]` (and other non-inline) annotated functions in Brillig mode (the default when debugging) by forcing the selection of the Brillig runtime when compiling them. For inline functions, during SSA generation the runtime is _not changed_ in order to allow some optimizations in the stdlib to still work. For example, [some constant assertions](https://github.com/noir-lang/noir/blob/9c6de4b25d318c6d211361dd62a112a9d2432c56/noir_stdlib/src/field.nr#L6) don't work unless the function is inlined.

When debugging in ACIR mode (with the `--acir-mode` CLI or `generateAcir: true` DAP options), the debugger now properly handles programs with multiple circuits and ACIR calls by:

1. keeping a stack of ACVMs and a record of the circuit being executed by each one, and
2. extending the debug location to also indicate in which circuit is the current opcode being executed

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by vezenovm** [comment]
Thank you for the PR @ggiraldez! I plan to look at this soon, just have been taken away by other tasks

File: tooling/debugger/src/context.rs [hunk 1]
```diff
@@ -15,56 +16,138 @@ use nargo::errors::{ExecutionError, Location};
...
+    #[error("Invalid debug location string: {0}")]
+    InvalidDebugLocationString(String),
+}
+
+impl std::str::FromStr for DebugLocation {
+    type Err = DebugLocationFromStrError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<_> = s.split(':').collect();
+
+        if parts.is_empty() || parts.len() > 2 {
```
Comment (vezenovm) [inline tooling/debugger/src/context.rs:53]: This feels like if could be cleaned up a little bit. We return an error here, panic inside of `parse_components`, and then do an `ok_or` that will never be hit because we only return `Some(..)` or panic.

File: tooling/debugger/src/context.rs [hunk 2]
```diff
@@ -15,56 +16,138 @@ use nargo::errors::{ExecutionError, Location};
...
+                    let opcode_location = OpcodeLocation::from_str(parts[1]).ok()?;
+                    Some(DebugLocation { circuit_id, opcode_location })
+                }
+                _ => unreachable!("`DebugLocation` has too many components"),
+            }
+        }
+
+        parse_components(&parts)
+            .ok_or(DebugLocationFromStrError::InvalidDebugLocationString(s.to_string()))
+    }
```
Comment (vezenovm) [inline tooling/debugger/src/context.rs:74]:
```suggestion
            match parts.len() {
                1 => {
                    let opcode_location = OpcodeLocation::from_str(parts[0]).ok()?;
                    Some(DebugLocation { circuit_id: 0, opcode_location })
                }
                2 => {
                    let circuit_id = parts[0].parse().ok()?;
                    let opcode_location = OpcodeLocation::from_str(parts[1]).ok()?;
                    Some(DebugLocation { circuit_id, opcode_location })
                }
                _ => return Err(DebugLocationFromStrError::InvalidDebugLocationString(s.to_string())),
            }
```
I think just this will work

File: tooling/debugger/src/context.rs [hunk 3]
```diff
@@ -15,56 +16,138 @@ use nargo::errors::{ExecutionError, Location};
...
     unconstrained_functions: &'a [BrilligBytecode<FieldElement>],
 
-    // Absolute (in terms of all the opcodes ACIR+Brillig) addresses of the ACIR
-    // opcodes with one additional entry for to indicate the last valid address.
-    acir_opcode_addresses: Vec<usize>,
+    // We define the address space as a contiguous space where all ACIR and
+    // Brillig opcodes from all circuits are laid out.
+    // The first element of the outer vector will contain the addresses at which
+    // all ACIR opcodes of the first circuit are. The second element will
+    // contain the second circuit and so on. Brillig functions should are
```
Comment (vezenovm) [inline tooling/debugger/src/context.rs:112]: This sentence isn't super clear, could you fix it up?

File: tooling/debugger/src/context.rs [hunk 4]
```diff
@@ -208,48 +305,56 @@ impl<'a, B: BlackBoxFunctionSolver<FieldElement>> DebugContext<'a, B> {
...
+        if address >= *self.acir_opcode_addresses.last()?.first()? {
             return None;
         }
-        let location = match self.acir_opcode_addresses.binary_search(&address) {
+        let circuit_id = match self
+            .acir_opcode_addresses
+            .binary_search_by(|addresses| addresses[0].cmp(&address))
+        {
+            Ok(found_index) => found_index,
+            Err(insert_index) => insert_index - 1,
```
Comment (vezenovm) [inline tooling/debugger/src/context.rs:341]: I'm surprised to see us use the insert index from an error. Don't we build all acir opcode addresses when we create the debugger context? Should this not be an internal debugger error?

File: tooling/debugger/src/context.rs [hunk 5]
```diff
@@ -15,56 +16,138 @@ use nargo::errors::{ExecutionError, Location};
...
     Done,
     Ok,
-    BreakpointReached(OpcodeLocation),
+    BreakpointReached(DebugLocation),
     Error(NargoError),
 }
 
+pub struct ExecutionFrame<'a, B: BlackBoxFunctionSolver<FieldElement>> {
+    circuit_id: u32,
+    acvm: ACVM<'a, FieldElement, B>,
```
Comment (vezenovm) [inline tooling/debugger/src/context.rs:87]: I'm wary that need to store the entire ACVM here? I only saw it really being used to fetch the instruction pointer and status. If there is something I missed could you elaborate?

File: tooling/debugger/src/context.rs [hunk 6]
```diff
@@ -15,56 +16,138 @@ use nargo::errors::{ExecutionError, Location};
...
+    // We define the address space as a contiguous space where all ACIR and
+    // Brillig opcodes from all circuits are laid out.
+    // The first element of the outer vector will contain the addresses at which
+    // all ACIR opcodes of the first circuit are. The second element will
+    // contain the second circuit and so on. Brillig functions should are
+    // expanded on each ACIR Brillig call opcode.
+    // At the end of each vector (both outer and inner) there will be an extra
+    // sentinel element with the address of the next opcode. This is only to
+    // make bounds checking and working with binary search easier.
+    acir_opcode_addresses: Vec<Vec<usize>>,
```
Comment (vezenovm) [inline tooling/debugger/src/context.rs:117]: It would be clearer to have this as its own type so that the address of the next opcode is simply a separate field rather than an extra sentinel element in the acir opcode list.

File: tooling/debugger/src/context.rs (hunk 4 above)
Comment (ggiraldez) [reply tooling/debugger/src/context.rs:341]: I see why this is confusing and definitely supports your suggestion about extracting the address mapping to a different type. We're using the insert index because we're doing a binary search of the addresses of the first opcode of each circuit only. This is to find which circuit the address belongs to.

And in the second binary search below this, it's also necessary because we're tracking the addresses of the ACIR opcodes only. Since the program is flattened for the purposes of computing the addresses, Brillig calls effectively "inline" all the Brillig opcodes in the same address space, so there may be holes in the nested vectors of addresses as well.

File: tooling/debugger/src/context.rs (hunk 5 above)
Comment (ggiraldez) [reply tooling/debugger/src/context.rs:87]: This is used to suspend the execution of an ACIR circuit while resolving an ACIR call, similar to how it's done [here](https://github.com/noir-lang/noir/blob/master/tooling/nargo/src/ops/execute.rs#L67). Only we cannot use recursion and the stack in this case, so we need to improvise our own stack.

An alternative would be to serialize the ACVM state and after the ACIR call is resolved, restore the parent ACVM with the serialized state. But that would require changes in the ACVM code itself.

File: tooling/debugger/src/context.rs (hunk 6 above)
Comment (ggiraldez) [reply tooling/debugger/src/context.rs:117]: The sentinel was initially added (before multiple circuits) to more easily handle the case when the last plus one address was being mapped to the opcode location. Without the sentinel, the function above would return a Brillig opcode location with the Brillig index beyond the unconstrained function's opcodes length. That was incorrect and problematic when handling the returned opcode location.

But we now introduced [a bounds check](https://github.com/manastech/noir/blob/feat%2F4824-multiple-acir-calls/tooling/debugger/src/context.rs#L333), which may as well check against a dedicated sentinel field.

File: tooling/debugger/src/context.rs (hunk 4 above)
Comment (vezenovm) [reply tooling/debugger/src/context.rs:341]: Ok thank you for elaborating. Yeah I do think this would be much easier to interpret if it was extracting into a type.

File: tooling/debugger/src/context.rs (hunk 5 above)
Comment (vezenovm) [reply tooling/debugger/src/context.rs:87]: Ah yeah I understand. I don't think we should change the ACVM interfaces for now and de/serializing the entire ACVM state is probably more overhead than just storing the ACVM stack in the way you have here. We can leave this as is.

File: compiler/noirc_frontend/src/hir/comptime/tests.rs [hunk 7]
```diff
@@ -31,14 +30,14 @@ fn interpret_expect_error(src: &str, func_namespace: Vec<String>) -> Interpreter
 
 #[test]
 fn interpreter_works() {
-    let program = "comptime fn main() -> pub Field { 3 }";
+    let program = "fn main() -> pub Field { 3 }";
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/hir/comptime/tests.rs:33]: I see lots of changes from master being reverted that are unrelated to this PR here. Bad merge?

**Comment by anaPerezGhiglia** [comment]
Hey @vezenovm! yes, it seems that somehow I messed it up on this merge [a322e9c](https://github.com/noir-lang/noir/pull/5051/commits/a322e9cd39804f681614367bdd318ed03d3f70f2)

I'll look into it and let you know once this is _actually_ ready to be reviewed. Sorry for the mess

**Comment by vezenovm** [comment]
> Hey @vezenovm! yes, it seems that somehow I messed it up on this merge [a322e9c](https://github.com/noir-lang/noir/pull/5051/commits/a322e9cd39804f681614367bdd318ed03d3f70f2)
> 
> I'll look into it and let you know once this is _actually_ ready to be reviewed. Sorry for the mess

No worries, apologies for the delay. Looking at https://github.com/manastech/noir/pull/3 it seems the comments have been addressed. Just ping once the merge is cleaned up!

**Comment by anaPerezGhiglia** [comment]
Hey @vezenovm I just updated the branch. I had to force-push it to remove the last merge commits. There were too many files to be addressed my manually solve the merge mess. The first part of the branch history remained the same. I just re-wrote the lasts commits. 

[Here is the comprare](https://github.com/noir-lang/noir/pull/5051/files/bf9461f84d5b0dd6ae1b46a4633795bb7e7760b2..d45f8e792ae1afee1592b7b31763febb5dbcb868) from your last formal review up to this point

File: tooling/debugger/src/context.rs [hunk 8]
```diff
@@ -15,56 +16,235 @@ use nargo::NargoError;
 use noirc_artifacts::debug::{DebugArtifact, StackFrame};
 use noirc_driver::DebugFile;
 
+use thiserror::Error;
+
 use std::collections::BTreeMap;
 use std::collections::{hash_set::Iter, HashSet};
 
+/// A Noir program is composed by
```
Comment (vezenovm) [inline tooling/debugger/src/context.rs:24]: Thank you for these comments. The map is much clearer now.

**Review by vezenovm — APPROVED** [review]

**Comment by vezenovm** [comment]
Let's get this in!
